rfcs/20200211-tf-types.md

```diff
@@ -0,0 +1,295 @@
+
+```
+>>> class SupportsTensor(Protocol):
+...   def __tf_tensor__(self):
+...     pass
+>>> def f(x: SupportsTensor):
+...   pass
+>>> obj = MyClass()
+>>> f(obj)  # passes static type checks
+```
+
+Ultimately, `TensorLike` is to become a union of the protocol type along with any other types supported through legacy mechanisms:
+
+```
+TensorLike = Union[List, Tuple, tf.Tensor, SupportsTensor, ...]
+```
+
+The `Protocol` type is only standardized in Python 3.8. Backports exist through [typing_extensions](https://github.com/python/typing/tree/master/typing_extensions), although they still don’t support Python 3.5. Therefore, typing annotations will only be supported in 3.6+, and complete support is only available in 3.8+.
+
+Note that `Protocol` subtypes require [@runtime_checkable](https://www.python.org/dev/peps/pep-0544/#runtime-checkable-decorator-and-narrowing-types-by-isinstance) in order to be compatible with `isinstance`. However, that degrades the performance of `isinstance` in a way similar to `abc.ABCMeta`. For that reason, TensorFlow internal logic is encouraged to use the the more direct `hasattr` test for structural type checks of this kind.
+
+Although this RFC proposes the deprecation of `register_tensor_conversion_function`, it does not set a timeline for removing it. It remains an open question whether interim support for type annotations should be added to `register_tensor_conversion_function`.
+
+### Support for `tf.function`'s `input_signature`
+
+The type system listed here can be expanded to allow input signatures using type annotations, see for instance [this thread](https://github.com/tensorflow/tensorflow/issues/31579).
+
+Presently, the [input_signature](https://www.tensorflow.org/api_docs/python/tf/function) optional mechanism uses [tf.TensorSpec](https://www.tensorflow.org/api_docs/python/tf/TensorSpec) to describe the function arguments:
+
+```
+>>> @function(input_signature=[TensorSpec([3], dtype=int32)])
+... def f(x):
+...   tf.print(x)
+>>> f(constant([1, 2, 3]))
+[1 2 3]
+>>> f(constant([1, 2]))  # Shape mismatch
+ValueError: Python inputs incompatible with input_signature
+>>> f(constant([1.0, 2.0, 3.0]))  # DType mismatch
+ValueError: Python inputs incompatible with input_signature
+```
+
+It is expected however that some or all of this information will be repeated by the function's type annotations. Type annotations may be generic, for example by only specifying a dtype:
+
+```
+>>> @function(input_signature=[TensorSpec([3], dtype=int32)])
+... def f(x: Tensor[int32]):
+...   ...
+```
+
+In such cases, `tf.function` is expected to verify that the type annotation matches the `input_signature`.
+
+In the long term, this RFC recommends that type annotations fully replace the `input_signature`, so far as the Python type annotation system allows it:
+
+```
+>>> class BatchSize(Dimension):
+...   value = 32
+>>> @function
+... def f(x: Tensor[Int32, Shape1D[BatchSize]]):
```

Comment:
What would be the recommended way to  type functions which accept multiple dtype? Would Union work? For example: `def f(x: Tensor[Union[tf.float32, tf.float64], Shape[32]]`.

Reply:
Although I haven't tested it, I'd expect Union to work because Tensor is a simple [Generic](https://docs.python.org/3/library/typing.html#typing.Generic) here. If that doesn't work, we could fall back to a less-powerful `AnyDType` identifier.